tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py

```diff
@@ -121,10 +121,73 @@ def tile_batch(t, multiplier, name=None):
     return nest.map_structure(lambda t_: _tile_batch(t_, multiplier), t)
 
 
+def gather_tree_from_array(t, parent_ids, sequence_length):
+  """Calculates the full beams for `TensorArray`s.
+
+  Args:
+    t: A `TensorArray` of size `max_time` that contains `Tensor`s of shape
+      `[batch_size, beam_width, s]` or `[batch_size * beam_width, s]` where
+      `s` is the depth shape.
+    parent_ids: The parent ids of shape `[max_time, batch_size, beam_width]`.
+    sequence_length: The sequence length of shape `[batch_size, beam_width]`.
+
+  Returns:
+    A `TensorArray` of the same size and type as `t` and where beams are sorted
+    in each `Tensor` according to `parent_ids`.
+  """
+  max_time = array_ops.shape(parent_ids)[0]
+  batch_size = array_ops.shape(parent_ids)[1]
+  beam_width = array_ops.shape(parent_ids)[2]
+
+  # Generate beam ids that will be reordered by gather_tree.
+  beam_ids = array_ops.expand_dims(
+      array_ops.expand_dims(math_ops.range(beam_width), 0), 0)
+  beam_ids = array_ops.tile(beam_ids, [max_time, batch_size, 1])
+
+  mask = array_ops.sequence_mask(
+      sequence_length, maxlen=max_time, dtype=dtypes.int32)
+  mask = array_ops.transpose(mask, perm=[2, 0, 1])
+
+  # Use beam_width + 1 to mark the end of beam.
+  masked_beam_ids = (beam_ids * mask) + (1 - mask) * (beam_width + 1)
+
+  max_sequence_lengths = math_ops.to_int32(
+      math_ops.reduce_max(sequence_length, axis=1))
+  sorted_beam_ids = beam_search_ops.gather_tree(
+      step_ids=masked_beam_ids,
+      parent_ids=parent_ids,
+      max_sequence_lengths=max_sequence_lengths,
+      end_token=beam_width + 1)
+
+  # For out of range steps, simply copy the same beam.
+  sorted_beam_ids = array_ops.where(
+      math_ops.cast(mask, dtypes.bool), x=sorted_beam_ids, y=beam_ids)
+
+  # Generate indices for gather_nd.
+  time_ind = array_ops.tile(array_ops.reshape(
+      math_ops.range(max_time), [-1, 1, 1]), [1, batch_size, beam_width])
+  batch_ind = array_ops.tile(array_ops.reshape(
+      math_ops.range(batch_size), [-1, 1, 1]), [1, max_time, beam_width])
+  batch_ind = array_ops.transpose(batch_ind, perm=[1, 0, 2])
+  indices = array_ops.stack([time_ind, batch_ind, sorted_beam_ids], -1)
+
+  # Gather from a tensor with collapsed additional dimensions.
+  gather_from = t.stack()
+  final_shape = array_ops.shape(gather_from)
+  gather_from = array_ops.reshape(
+      gather_from, [max_time, batch_size, beam_width, -1])
+  ordered = array_ops.gather_nd(gather_from, indices)
+  ordered = array_ops.reshape(ordered, final_shape)
+
+  # Return the result as a TensorArray.
+  ordered_ta = tensor_array_ops.TensorArray(
```

Comment:
i'm wondering if we need to go through this unstack.  maybe just say in the initializer docstring "if the TensorArray can be reordered, it will be reordered and returned back in reordered and stacked form (i.e., as a regular tensor)".  TensorArrays that cannot be reordered will be returned in non-stacked form.

?

Reply:
We could do that if it is properly documented but it would make the API a bit less consistent. Taking `alignement_history` as an example, the user would first expect the same output type from a `BeamSearchDecoder` or a `BasicDecoder`.

What about stacking all TensorArrays in the cell state in `finalize`? It seems like that is what users do anyway.

Let me know what you think is best.

Reply:
thank you @guillaumekln +1 for the non-stacked form. eventually i still need to stack it up. unstack here and stack in finalize seems redundant. i would prefer as less unnecessary oprs as possible.  it is up to ebrevdo and you though.